### DISCUSSION TITLE
Google cloud speech to text fails with exception : Unavailable

### DISCUSSION THREAD

**Discussion by sychopath** [googleapis/google-cloud-cpp#9771] [Q&A] [answered]
I am trying to use google-cloud-cpp speech to text apis. I am able to run sample codes using as a standalone application by building a shared object(.so) file have google stt code and calling same functions through an executable.
When I am using the same set of library to call apis inside my application plugin(Unimrcp), I get the below exception(status):

**UNAVAILABLE: Error in non-idempotent operation Recognize: failed to connect to all addresses; last error: UNKNOWN: Network is unreachable**

System information : I am using a gcp server with Ubuntu 18.04.6 LTS.  Built a sample app and unimrcp application using Makefiles.
I have saved google credentials through environment variable. Also, have verified that the request is not reaching my api server as there is no log.

There should not be any network issues as I am to reach apis through another executable. Moreover, I am using another set of plugin with my Unimrcp service for azure stt, which works fine.

I have also captured data packets using wireshark, a snapshot of that is attached.

<img width="1635" alt="Screenshot 2022-09-04 at 3 15 34 PM" src="https://user-images.githubusercontent.com/5571122/188307526-284c2fc7-6f32-446b-94d9-ba910fa432bf.png">

**Comment by coryan** [comment]
> I am trying to use google-cloud-cpp speech to text apis.

Do you know what version you are using? Likewise for gRPC, protobuf, Abseil, and OpenSSL?

> When I am using the same set of library to call apis inside my application plugin(Unimrcp), I get the below exception(status):
> 
> **UNAVAILABLE: Error in non-idempotent operation Recognize: failed to connect to all addresses; last error: UNKNOWN: Network is unreachable**

Is it possible to run this code with the following environment variables?  That will add some logging that may help: `GOOGLE_CLOUD_CPP_ENABLE_CLOG=yes` and `GOOGLE_CLOUD_CPP_ENABLE_TRACING=rpc,rpc-streams`.


> Also, have verified that the request is not reaching my api server as there is no log.

I am not sure what is your API server.  Nor what request you are referring to.  But I take it that this is an indication that the TTS request failed.

> There should not be any network issues as I am to reach apis through another executable. Moreover, I am using another set of plugin with my Unimrcp service for azure stt, which works fine.

When you use `google-cloud-cpp` as part of the plugin: is this plugin dynamically loaded?

**Reply by sychopath** [reply]
Thank you for your response.

I have taken the release versions for all the dependencies including the ones you have mentioned. I checked the Cmake version, which is 3.20.0.

> Is it possible to run this code with the following environment variables? That will add some logging that may help: GOOGLE_CLOUD_CPP_ENABLE_CLOG=yes and GOOGLE_CLOUD_CPP_ENABLE_TRACING=rpc,rpc-streams.

I have added these environment variables and can see logs on console while running my unimrcp server. But these logs do not add to our knowledge anything we do not know :

> 2022-09-05T14:44:19.162925307Z [INFO] <140114416088832> Enabled logging for gRPC calls (/root/vcpkg/buildtrees/google-cloud-cpp/src/v2.1.0-54fb822ad3.clean/google/cloud/speech/internal/speech_stub_factory.cc:53)

> 2022-09-05T14:44:19.163618141Z [DEBUG] <140114416088832> Recognize() << google.cloud.speech.v1.RecognizeRequest { config { language_code: "en-US" } audio { uri: "gs://cloud-samples-data/speech/hello.wav" } } (/root/vcpkg/buildtrees/google-cloud-cpp/src/v2.1.0-54fb822ad3.clean/google/cloud/internal/log_wrapper.h:91)

> 2022-09-05T14:44:20.225469631Z [DEBUG] <140114416088832> Recognize() >> status=UNAVAILABLE: failed to connect to all addresses; last error: UNKNOWN: Network is unreachable (/root/vcpkg/buildtrees/google-cloud-cpp/src/v2.1.0-54fb822ad3.clean/google/cloud/internal/log_wrapper.h:94)

I suppose we have already seen this previously.

> When you use google-cloud-cpp as part of the plugin: is this plugin dynamically loaded?

Yes, this gets loaded dynamically.

With respect to **server** on my previous post, I was referring to google console dashboard where I can view requests log and did not find any there. 

Thanks again !!

**Reply by coryan** [reply]
> Thank you for your response.
> 
> I have taken the release versions for all the dependencies including the ones you have mentioned. I checked the Cmake version, which is 3.20.0.

Thanks.

> > Is it possible to run this code with the following environment variables? That will add some logging that may help: GOOGLE_CLOUD_CPP_ENABLE_CLOG=yes and GOOGLE_CLOUD_CPP_ENABLE_TRACING=rpc,rpc-streams.
> 
> I have added these environment variables and can see logs on console while running my unimrcp server. But these logs do not add to our knowledge anything we do not know :

Yes, not a very revealing log.  It tells me a couple of things, mostly by what is not in the log.  For example, this request did not go through the retry loop.  If we find that this is a transient error we can fix that (there are reasons why it is not the default).  It also shows that this is not an authentication problem, and that you compiled `google-cloud-cpp` using `vcpkg`.

> 
> > 2022-09-05T14:44:19.162925307Z [INFO] <140114416088832> Enabled logging for gRPC calls (/root/vcpkg/buildtrees/google-cloud-cpp/src/v2.1.0-54fb822ad3.clean/google/cloud/speech/internal/speech_stub_factory.cc:53)
> 
> > 2022-09-05T14:44:19.163618141Z [DEBUG] <140114416088832> Recognize() << google.cloud.speech.v1.RecognizeRequest { config { language_code: "en-US" } audio { uri: "gs://cloud-samples-data/speech/hello.wav" } } (/root/vcpkg/buildtrees/google-cloud-cpp/src/v2.1.0-54fb822ad3.clean/google/cloud/internal/log_wrapper.h:91)
> 
> > 2022-09-05T14:44:20.225469631Z [DEBUG] <140114416088832> Recognize() >> status=UNAVAILABLE: failed to connect to all addresses; last error: UNKNOWN: Network is unreachable (/root/vcpkg/buildtrees/google-cloud-cpp/src/v2.1.0-54fb822ad3.clean/google/cloud/internal/log_wrapper.h:94)

This error comes from gRPC:

https://github.com/grpc/grpc/blob/fed6c180e96f6beb58f79c37cb10b197021739c0/src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc#L425-L427

I think the next debugging step is to enable gRPC logging.  I am less confident parsing the gRPC output, but I will give it a try.  Do you mind setting `GRPC_VERBOSITY=debug` and  `GRPC_TRACE=tcp,http,api`  as described in:

https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md

> > When you use google-cloud-cpp as part of the plugin: is this plugin dynamically loaded?
> 
> Yes, this gets loaded dynamically.

Okay.  This is unlikely, but some C++ libraries do not work well when dynamically loaded (specially if you use `RTLD_LOCAL` and then load multiple `.so` libraries).  I think we can ignore that problem for now, but if we eliminate other problems we should revisit this.

> With respect to **server** on my previous post, I was referring to google console dashboard where I can view requests log and did not find any there.

Okay.  I would not expect to see anything in the console with such an error.  The request never made it to the portions of GCP that would be able to log this.

**Reply by sychopath** [reply]
[grpc_logs.txt](https://github.com/googleapis/google-cloud-cpp/files/9503812/grpc_logs.txt)

Thank you for your suggestions. I have added grpc environment variables .

> I think the next debugging step is to enable gRPC logging. I am less confident parsing the gRPC output, but I will give it a try. Do you mind setting GRPC_VERBOSITY=debug and GRPC_TRACE=tcp,http,api as described in:

> https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md

I have attached the grpc logs from console. I can see an error related to security handshake :

**Security handshake failed: UNKNOWN:Peer name speech.googleapis.com is not in peer certificate**

Not sure why the same set of code/library works fine at one place and fails at other.

**Reply by coryan** [reply]
> 
> Thank you for your suggestions. I have added grpc environment variables .

Wow, that is a lot of logging. Sorry you had to wade through that much.

> I have attached the grpc logs from console. I can see an error related to security handshake :
> 
> **Security handshake failed: UNKNOWN:Peer name speech.googleapis.com is not in peer certificate**
> 
> Not sure why the same set of code/library works fine at one place and fails at other.

I am not sure either.  There are a couple of hypothesis that come to mind:

- The environment where you have deployed this does not have the `ca-certificates` package (or a similar name) installed.  That sometimes happens when creating Docker images, but you said the Azure equivalent worked, so that seems unlikely.
- If the OpenSSL library is linked dynamically you may be using a different OpenSSL library in one environment vs. the other.  It may be that the two OpenSSL versions behave differently, e.g., expect to find the root certificates in different locations. From your previous comments you are using `vcpkg` to compile `google-cloud-cpp` and its dependencies.
  - Can you let us know which triplet you used to compile the libraries?
  - Does that triplet produce shared libraries or static libraries?
- (unlikely) Maybe the environment where the plugin is loaded has limited access to the file system? Maybe via some kind of container or via some kind of LD_PRELOAD?

As a workaround (or maybe to test some of these hypothesis).   You can try setting `GRPC_DEFAULT_SSL_ROOTS_FILE_PATH` and point this to a `roots.pem` file that you create yourself. For example, you can download https://pki.google.com/roots.pem and point the environment variable to it.

**Reply by sychopath** [reply]
> The environment where you have deployed this does not have the ca-certificates package (or a similar name) installed. That sometimes happens when creating Docker images, but you said the Azure equivalent worked, so that seems unlikely.

I am loading my plugin inside a service. But there could be a possibility that my plugin is not able to access my environment variables.

> If the OpenSSL library is linked dynamically you may be using a different OpenSSL library in one environment vs. the other.

I have fixed this by updating openssl to 3.0.5 version, had an older version installed on my system.

> From your previous comments you are using vcpkg to compile google-cloud-cpp and its dependencies.
Can you let us know which triplet you used to compile the libraries?
Does that triplet produce shared libraries or static libraries?

I have installed/downloaded all the libraries through vcpkg. I see only static library files(.a) installed on my system. I have followed the same instruction mentioned on github.

https://github.com/googleapis/google-cloud-cpp
https://github.com/googleapis/google-cloud-cpp/tree/main/google/cloud/speech/quickstart

> Maybe the environment where the plugin is loaded has limited access to the file system? Maybe via some kind of container or via some kind of LD_PRELOAD?

Not sure, but my plugin gets loaded inside a service or even from executable and I see no difference. 

> As a workaround (or maybe to test some of these hypothesis). You can try setting GRPC_DEFAULT_SSL_ROOTS_FILE_PATH and point this to a roots.pem file that you create yourself. For example, you can download https://pki.google.com/roots.pem and point the environment variable to it.

I downloaded the certificate from the link and have set the environment variable with its path but I don't see any differences in the result from standalone app or plugin execution. 
I looked into the grpc logs from my working app(executable) to see the contents of certificate exchange but don't see the data referred by new pem file.  I am new to grpc and not sure if I understood your point correctly. 

Thanks again, Really appreciate your help here @coryan !!

**Reply by coryan** [reply]
I should start by saying that I am not sure what the problem is, and you should not put too much weight on what I say.  I may be completely off-base and apologize if I waste your time.

I think at this point I really need to understand your link line to create the plugin `.so` file.

> I am loading my plugin inside a service. But there could be a possibility that my plugin is not able to access my environment variables.

I do not recall if you mentioned this: do you use the same operating system and distribution to develop and deploy your service and the plugin?  OpenSSL figures out where to find the root certificates at compile-time, but different distributions use different paths for the root certificates.  If you were (for example) developing on a Fedora or RedHat system the root certificates are at `/etc/pki/tls/certs/ca-bundle.crt`, if you then deploy to an Ubuntu or Debian system they would be at `/etc/ssl/certs/ca-certificates.crt`.  An OpenSSL version compiled on RedHat/Fedora could not find the certificates if deployed to Ubuntu or Debian.

> 
> > If the OpenSSL library is linked dynamically you may be using a different OpenSSL library in one environment vs. the other.
> 
> I have fixed this by updating openssl to 3.0.5 version, had an older version installed on my system.

I take it that this did not fix the problem, you just made sure that inconsistent OpenSSL versions were not the problem.  Did I read that right?

> I have installed/downloaded all the libraries through vcpkg. I see only static library files(.a) installed on my system. I have followed the same instruction mentioned on github.
> 
> https://github.com/googleapis/google-cloud-cpp https://github.com/googleapis/google-cloud-cpp/tree/main/google/cloud/speech/quickstart

That all sounds good.  Can you share how you build the plugin?  In particular, how do you invoke the compiler (or linker) to create the plugin `.so`?

> 
> > Maybe the environment where the plugin is loaded has limited access to the file system? Maybe via some kind of container or via some kind of LD_PRELOAD?
> 
> Not sure, but my plugin gets loaded inside a service or even from executable and I see no difference.

Oh... I think I missed something.  I thought this was the situation:

1. Running the quickstart works
1. Running your plugin in a standalone application works
1. Running your plugin in the service fails

I think you are saying that it does not work in the second case either, I had missed that.  🤔 

> I downloaded the certificate from the link and have set the environment variable with its path but I don't see any differences in the result from standalone app or plugin execution. I looked into the grpc logs from my working app(executable) to see the contents of certificate exchange but don't see the data referred by new pem file. I am new to grpc and not sure if I understood your point correctly.

Hmmm... that is very odd.  gRPC is pretty good at using the environment variable.  Is there any way you can add some debugging code to your plugin to show if this (and any other) environment variables are set?  Maybe print the output (if not `null`) of `std::getenv("...")`?

**Reply by sychopath** [reply]
> I should start by saying that I am not sure what the problem is, and you should not put too much weight on what I say. I may be completely off-base and apologize if I waste your time.

Thanks @coryan  for your response again. You are the best hope I have for solving this issue  :)

Below is the repository of the application I am extending on. The Makefile.am will give you the gist of how the plugins get loaded dynamically once I start my unimrcp service.

[https://github.com/mradmila/unimrcp](url)


> I do not recall if you mentioned this: do you use the same operating system and distribution to develop and deploy your service and the plugin? 
Yes, everything is happening on the same system, ie, Ubuntu 18.04

> I take it that this did not fix the problem, you just made sure that inconsistent OpenSSL versions were not the problem. Did I read that right?

I saw an older version of openssl and looked into vcpkg versions. Found vcpkg was referring to 3.0.5, so updated my system's version as well. I am not sure if this was of any use, just covering all cases.

> Running the quickstart works
Yes

> Running your plugin in a standalone application works

Yes, my code works fine inside a standalone application.

> Running your plugin in the service fails

Exactly, it fails inside the plugin.

> Hmmm... that is very odd. gRPC is pretty good at using the environment variable. Is there any way you can add some debugging code to your plugin to show if this (and any other) environment variables are set? Maybe print the output (if not null) of std::getenv("...")?

Got below logs today. Printed env variables inside the plugin and looks like it is getting the values :

2022-09-12 07:46:05:259939 [INFO]   *******************Demo Recog Engine : checking env variables -----------------  :  0

******************* print env variable ------cout ------/home/ubuntu/roots.pem

*******************  print env variable ------cout ------/home/ubuntu/service-account-file.json

**Reply by coryan** [reply]
> Below is the repository of the application I am extending on. The Makefile.am will give you the gist of how the plugins get loaded dynamically once I start my unimrcp service.
> 
> [https://github.com/mradmila/unimrcp](url)

As best as I can tell, you are trying to write a plugin similar to this example:

https://github.com/mradmila/unimrcp/tree/master/plugins/demo-recog

it would be useful to know the *exact* command use to create your plugin `.so`.  Or at least how does this `.so` link `google-cloud-cpp` and its dependencies (e.g. the `-lgoogle_cloud_cpp_speech` option and anything after that).  It would also be useful to know if this plugin links `google-cloud-cpp` and its dependencies as static libraries or as shared libraries.

This is starting to remind me of:

https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/846

It also reminds me of:

https://github.com/abseil/abseil-cpp/issues/1230

nit: your link was broken, but it is easy to cut&paste and see what you meant.

> > Running the quickstart works
>  Yes
> 
> > Running your plugin in a standalone application works
> 
> Yes, my code works fine inside a standalone application.
> 
> > Running your plugin in the service fails
> 
> Exactly, it fails inside the plugin.

Okay. This narrows down the problem to either (1) the environment (including file system access and environment variables) inside the plugin, (2) maybe some container issues, if your server is running in a container, (3) some weirdness with dynamically loading the code.  Given your testing of the environment variables inside the plugin, I am more inclined to believe it is (3).  That *could* be my bias, because I have had trouble with dynamic loading in the past.  Seems worthwhile to rule out at least.

This is going to be difficult to troubleshoot.  We need to know a few things:

- What is the output of `ldd ${my-plugin-name}.so`?  What dependencies of the plugin are dynamically loaded?
- What are the options used to load the plugin?  Probably using [LD_DEBUG](https://bnikolic.co.uk/blog/linux-ld-debug.html) with the right values can give us that?
- How are symbols resolved when this library is loaded?  That matters for complicated reasons explained here: https://akkadia.org/drepper/dsohowto.pdf

**Comment by sychopath** [comment, marked as answer]
Thanks for your response @coryan !!

Infact below is the working link we are using for microsoft stt and have plans to add google stt to it.

[https://github.com/mradmila/unimrcp/tree/master/plugins/ms-recog](url)


> It would also be useful to know if this plugin links google-cloud-cpp and its dependencies as static libraries or as shared libraries.

I am attaching my Makfile.am which I have used to build by cpp code. 

[Make-gcp-recog.txt](https://github.com/googleapis/google-cloud-cpp/files/9582008/Make-gcp-recog.txt)

> nit: your link was broken, but it is easy to cut&paste and see what you meant.

my bad.

> Okay. This narrows down the problem to either (1) the environment (including file system access and environment variables) inside the plugin

I was able to print the environment variables value from code, not sure if this can still be a problem. 

> (2) maybe some container issues, if your server is running in a container

No, simple service or executable 

> (3) some weirdness with dynamically loading the code. Given your testing of the environment variables inside the plugin, I am more inclined to believe it is

very high probability


> What is the output of ldd ${my-plugin-name}.so? What dependencies of the plugin are dynamically loaded?

ldd gsrecog.so.0
	linux-vdso.so.1 (0x00007ffdc3f30000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007ff8e9d43000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ff8e9b24000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ff8e979b000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ff8e93fd000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ff8e900c000)
	/lib64/ld-linux-x86-64.so.2 (0x00007ff8eb151000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ff8e8df4000)

**Reply by coryan** [reply]
> Infact below is the working link we are using for microsoft stt and have plans to add google stt to it.
> 
> [https://github.com/mradmila/unimrcp/tree/master/plugins/ms-recog](url)
> 
> > It would also be useful to know if this plugin links google-cloud-cpp and its dependencies as static libraries or as shared libraries.
> 
> I am attaching my Makfile.am which I have used to build by cpp code.
> 
> [Make-gcp-recog.txt](https://github.com/googleapis/google-cloud-cpp/files/9582008/Make-gcp-recog.txt)

Thanks.  One thing seems curious. The `Makefile.am` for `ms-recog` reads:

```Makefile
msrecog_la_LDFLAGS         = $(UNIMRCP_PLUGIN_OPTS) \
                             -L$(SPEECHSDK_ROOT)/lib/x64 \
                             -lpthread -lMicrosoft.CognitiveServices.Speech.core
```

While your `Makefile.am` for `gcp-recog` reads:

```
gsrecog_la_LDFLAGS         = $(UNIMRCP_PLUGIN_OPTS) \
                             $(shell pkg-config google_cloud_cpp_speech --libs-only-L)
```

That is, you are only providing the `-L` flags, but seems like you want both `-L` and `-l` flags.  So this should be:

```
gsrecog_la_LDFLAGS         = $(UNIMRCP_PLUGIN_OPTS) \
                             $(shell pkg-config google_cloud_cpp_speech) -lupb_fastdecode -lupb_utf8_range
```

I doubt this is the cause of the problems, because your plugin would simply not link in that case.

It has been a long time since I used automake / autotools, but the package referenced uses C as the default language:

https://github.com/mradmila/unimrcp/blob/f3950b696456833ad316c5d90d7724969885ae29/configure.ac#L47

I wonder if the plugin needs to be explicitly linked using the CXX tag:

https://www.gnu.org/software/libtool/manual/html_node/Tags.html

That is pure speculation.


> > Okay. This narrows down the problem to either (1) the environment (including file system access and environment variables) inside the plugin
> 
> I was able to print the environment variables value from code, not sure if this can still be a problem.

I am thinking more obscure problems, such as the service using `LD_PRELOAD` tricks to limit the files that a plugin can open.  Very unlikely though.


> > What is the output of ldd ${my-plugin-name}.so? What dependencies of the plugin are dynamically loaded?
> 
> ldd gsrecog.so.0 linux-vdso.so.1 (0x00007ffdc3f30000) librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007ff8e9d43000) libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ff8e9b24000) libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ff8e979b000) libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ff8e93fd000) libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ff8e900c000) /lib64/ld-linux-x86-64.so.2 (0x00007ff8eb151000) libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ff8e8df4000)

Looks like all the `google_cloud_cpp*` and `grpc*` libraries are statically linked into the `.so`.  That is probably the most reliable way to make this work.

At this point the only idea I have is to try to reproduce the problem with a smaller example.  Can you reduce this to a self-contained program (maybe a single `.so` that gets loaded by a single `main()`)?  I can try to do so myself, but I doubt I will have time today.

**Reply by coryan** [reply]
Based on your description I created a program that runs the `speech` quickstart via a dynamically-loaded plugin.  It works for me, which only means we need to identify what did I do differently vs. what happens in your application.

I will need your help with that. Things to look for:

- Make sure the plugin is linked with the C++ compiler (e.g. `g++ --shared`).  The C compiler (or `ld`) will not invoke the global initializers when the plugin is loaded.
- Make sure the `dlopen()` flags I used match what the application uses.  I am not suggesting the flags I used are the only right ones, just that they are important, and we need to know if there is a difference.  Using `LD_DEBUG` (links above) can help you find those.

[repro-9771.tar.gz](https://github.com/googleapis/google-cloud-cpp/files/9586248/repro-9771.tar.gz)

**Reply by sychopath** [reply]
Hey @coryan 

> At this point the only idea I have is to try to reproduce the problem with a smaller example. Can you reduce this to a self-contained program (maybe a single .so that gets loaded by a single main())?


Thanks for trying this out at your end. Actually sample code works fine for me as well. I have tried doing the same thing at my end before raising this question. May be it was not very much clear from my first post of the discussion : 

> I am able to run sample codes using as a standalone application by building a shared object(.so) file have google stt code and calling same functions through an executable.

I did build a similar code, with a Makefile which works fine. I can share the grpc logs from my end if it is of any use.

I will take a look into your Makefile and figure out differences. Thanks for sharing !!

**Reply by sychopath** [reply]
Hey @coryan ,

> Thanks. One thing seems curious. The Makefile.am for ms-recog reads:

I have made changes to my Makefile.am, as per your suggestions. Thanks for pointing them out.
[improved_Makefile.txt](https://github.com/googleapis/google-cloud-cpp/files/9592060/improved_Makefile.txt)

> Using LD_DEBUG (links above) can help you find those.

Ok, I am not so sure how to find those flags using LD_DEBUG, but I am attaching the console log where I have tried 3 different values of LD_DEBUG.

[ld-all.zip](https://github.com/googleapis/google-cloud-cpp/files/9592070/ld-all.zip)
[LD_DEBUG-libs.txt](https://github.com/googleapis/google-cloud-cpp/files/9592064/LD_DEBUG-libs.txt)
[LD_DEBUG-files.txt](https://github.com/googleapis/google-cloud-cpp/files/9592067/LD_DEBUG-files.txt)

I need some help here from you again. I think the flag we are referring to is "rtld_global" Can you please confirm this if I am correct ? 
`LD_DEBUG =all` produces a huge log, but this is only where I see some data related to RTLD flags. 



Thanks again !!

**Reply by sychopath** [reply]
> Based on your description I created a program that runs the speech quickstart via a dynamically-loaded plugin. It works for me, which only means we need to identify what did I do differently vs. what happens in your application.

I tried running your application at my end and it worked fine. So I tried changing the flags here :

From :
`plugin = dlopen("/usr/src/repro-9771/plugin.so", RTLD_LOCAL | RTLD_LAZY);`

to :

`plugin = dlopen("/usr/src/repro-9771/plugin.so", RTLD_GLOBAL);`

I see an error at my end.

```
root@jioeva-cas-dev-6:/usr/src/repro-9771# ./main
Error opening plugin
```

Can you please confirm this behaviour ?

**Reply by coryan** [reply]
**TL;DR;** We need confirm:

- The `Makefile.am` is linking the plugin as a C++ shared object.
- What flags is `unimrcp` using to `dlopen()` the plugin, then we can use the same flags in a repro.
- We need to change the repro to link the system's OpenSSL library, this may be causing the wrong symbols to be resolved when loading the plugin.

----

Let me summarize the situation:

- **Case 1** A C++ program statically linked against `libgoogle_cloud_cpp_speech` works.
- **Case 2** A C++ program dynamically linked against `libgoogle_cloud_cpp_speech` (i.e. one where `ldd program` includes `libgoogle_cloud_cpp_speech`) works.
- **Case 3** A C program that dynamically loads (via `dlopen()`) a `.so` which includes `libgoogle_cloud_cpp_speech` works.
- **Case 4** Creating a plugin with the `unimrcp` build infrastructure and loading it into a `unimrcp` server does **not** work.

FWIW, when you said:

> I am able to run sample codes using as a standalone application by building a shared object(.so) file have google stt code and calling same functions through an executable.

I thought you were referring to "Case 2".  Regardless, since "Case 3" works, I think we need to concentrate on figuring out what are the differences between "Case 4" and "Case 3".  Does that make sense?

> I have made changes to my Makefile.am, as per your suggestions. Thanks for pointing them out. [improved_Makefile.txt](https://github.com/googleapis/google-cloud-cpp/files/9592060/improved_Makefile.txt)

I am not sure, but I think you are missing one more change.  As I noted above, this may be linking that `.so` as a **plan C** library, you may need to tell `libtool` to link it as a C++ program.  I have not used automake and libtool in a long time, and I do not recall the details of how to make the change I am proposing.  Maybe this link can help you:

https://www.gnu.org/software/libtool/manual/html_node/Tags.html

To troubleshoot this further, I would need the *exact* command that creates the `.so` file.  I am not sure how to do that, you may need to learn more about how `automake` and `libtool` debugging works.

> I need some help here from you again. I think the flag we are referring to is "rtld_global" Can you please confirm this if I am correct ? 

Yes, I am interested in the `RTLD_GLOBAL` flag vs. whatever else the application uses. 

> `LD_DEBUG =all` produces a huge log, but this is only where I see some data related to RTLD flags.

I expected LD_DEBUG to print the relevant info. Evidently I was wrong.  This seems to work for me:

```
ltrace -o ltrace.log main
grep dlopen ltrace.log
```

Can you use something similar to find the flags for `dlopen()`?

With that said.  I just noticed this in your output:

```
binding file /usr/local/unimrcp/plugin/gsrecog.so [0] to /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 [0]: normal symbol `X509_get_version'
```

That means the symbol resolution for an OpenSSL symbol used the symbols linked against the main program, not the symbols in `gsrecog.so`.  That seems wrong to me, as you are probably using different OpenSSL versions in the plugin vs. the main program.

----

> plugin = dlopen("/usr/src/repro-9771/plugin.so", RTLD_GLOBAL);
> I see an error at my end.
> Can you please confirm this behaviour ?

Confirmed.

**Reply by coryan** [reply]
I think the two versions of OpenSSL are the problem.  I changed `main` in my repro to use the system's OpenSSL library and got a crash.  I believe that using two versions and having the plugin resolve the symbols to either version is bound to produce problems.  Better to use a consistent version.

I have modified my repro to have two Dockerfiles.  `vcpkg.Dockerfile` results in crashes, `local.Dockerfile` works correctly.

[repro-9771.tar.gz](https://github.com/googleapis/google-cloud-cpp/files/9592628/repro-9771.tar.gz)

**Reply by sychopath** [reply]
Hey @coryan ,

> I think the two versions of OpenSSL are the problem.

I see 'X509_get_version' symbol is being searched in libcrypto present in /usr/lib while it should have used gsrecog.so to find that. I found a reference to openssl and crypto inside my libunimrcpserver.la file. I think while setting things up, I might have downloaded these files. I tried uninstalling openssl, but these files still exists. 

[https://github.com/unispeech/unimrcp/tree/master/platforms/libunimrcp-server](url) 


> I have modified my repro to have two Dockerfiles. vcpkg.Dockerfile results in crashes, `local.Dockerfile` works correctly.

Thanks a lot for sharing this.  I see an env variable declaration pointing vcpkg in your `local.Dockerfile`, but if I understand correctly, you do not intend to use vcpkg to build. You are building each component separately (re2, grpc, abseil,...).
Please correct me if I have mis-understood. 

Not sure if it will make a a difference but at this point I would also like your suggestion if I should proceed using vcpkg or should I install each component  ?

**Reply by coryan** [reply]
> Hey @coryan ,
> 
> > I think the two versions of OpenSSL are the problem.
> 
> I see 'X509_get_version' symbol is being searched in libcrypto present in /usr/lib while it should have used gsrecog.so to find that. I found a reference to openssl and crypto inside my libunimrcpserver.la file. I think while setting things up, I might have downloaded these files. I tried uninstalling openssl, but these files still exists.
> 

I am not sure if there is a question there.  I suspect `unimrcp` always needs to link `libcrypto.so`, so we better make 

> 
> > I have modified my repro to have two Dockerfiles. vcpkg.Dockerfile results in crashes, `local.Dockerfile` works correctly.
> 
> Thanks a lot for sharing this. I see an env variable declaration pointing vcpkg in your `local.Dockerfile`, but if I understand correctly, you do not intend to use vcpkg to build. You are building each component separately (re2, grpc, abseil,...). Please correct me if I have mis-understood.

That was a mistake, but had no effect.  You are correct, the intent was to build each component.  I have updated the repro:

[repro-9771.tar.gz](https://github.com/googleapis/google-cloud-cpp/files/9594285/repro-9771.tar.gz)


> 
> Not sure if it will make a a difference but at this point I would also like your suggestion if I should proceed using vcpkg or should I install each component ?

I am going to be a bit less directive, but hopefully more general:  you should make sure that `gsrecog.so` links against the same versions of any shared libraries used by your server.  So start by listing what shared objects are used in the server (e.g. `ldd my-server`), then set up your build environment to use the same version of any libraries you find there.  I think it would be easier to achieve this by installing each component, but you may find a different / better solution.

BTW, I don't think we have proven that the OpenSSL version problems are exactly what is causing the problems you described at the beginning of this discussion.  I think it is certainly **a** problem, and I suspect it is likely to be **the** problem, but I won't be surprised if there are other issues lurking here.

**Reply by sychopath** [reply]
Hey @coryan ,

I have built each component individually and everything works smoothly. There could be multiple problems with the approach I was taking while building with vcpkg. 

Thanks for helping me here !!
